### PR TITLE
add parts of master`s thesis as documentation

### PR DESCRIPTION
**Pull request by BenjaminW3**
Extract parts of the Master`s-Thesis into a in-repo documentation and convert it to markdown (see #270).

### CODE REVIEW

**Comment by BenjaminW3** [comment]
A big part of my master`s thesis has been reformatted into markdown and added as documentation. This is not meant to be perfect but it is far better then the current state of zero documentation!

File: doc/markdown/user/Abstraction.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+Multiple levels of caches, from small and very fast ones to very large and slower ones exploit temporal locality by keeping recently referenced data as close to the actual processing units as possible.
+Spatial locality in the main memory is also important for caches because they are usually divided into multiple lines that can only be exchanged one cache line at a time.
+If one data element is loaded and cached, it is highly likely that nearby elements are also cached.
+If the pixels of an image are stored row wise but are read out column wise, the spatial locality assumption of many CPUs is violated and the performance suffers.
+GPUs on the other hand do not have a large caching hierarchy but allow explicit access to a fast memory shared across multiple cores.
+Therefore, the best way to process individual data elements of a data parallel task is dependent on the data structure as well as the underlying hardware.
+%Even if an algorithm itself does not require the usage of shared data, it is often beneficial to utilize it, because the hardware is optimized for
```
Comment (psychocoderHPC) [inline doc/markdown/user/Abstraction.md]: Please remove the `%`

File: doc/markdown/user/abstraction/Element.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+To use the maximum available computing power of, for example, a modern x86 processor, the computation has to utilize the SIMD vector registers.
+Many current architectures support issuing a single instruction that can be applied to multiple data elements in parallel.
+
+The original x86 instruction set architecture did not support SIMD instructions but has been enhanced with MMX (64 bit width registers), SSE (128 bit width registers), AVX (256 bit width registers) and AVX-512 (512 bit width registers) extensions.
+In varying degree, they allow to process multiple 32 bit and 64 bit floating point numbers as well as 8, 16, 32 and 64 bit signed and unsigned integers.
+
+*CUDA* capable GPUs do not have vector registers where multiple values of type `float` or `double` can be manipulated by one instruction.
+Nevertheless, newer *CUDA* capable devices implement basic SIMD instructions on pairs of 16 bit values and quads of 8-bit values. 
+They are described in the documentation of the [PTX instruction set architecture](http://docs.nvidia.com/cuda/pdf/ptx_isa_3.2.pdf) chapter 8.7.13 but are only of any use in very special problem domains, for example for deep learning.
```
Comment (psychocoderHPC) [inline doc/markdown/user/abstraction/Element.md]: Maybe we should link to [this](http://docs.nvidia.com/cuda/parallel-thread-execution/index.html) online documenation which points always to the newest ISA version

File: doc/markdown/user/Introduction.md [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+Since 2012 no news were published and nothing could be found in the yearly release notes of the *PGI* compiler suite.
+
+
+### [GPU Ocelot](http://gpuocelot.gatech.edu/)
+<!--- https://github.com/gtcasl/gpuocelot --->
+is an open-source dynamic JIT compilation framework.
+It allows to execute native *CUDA* binaries by dynamically translating the *NVIDIA PTX* virtual instruction set architecture to other instruction sets.
+It supports *NVIDIA* and *AMD* GPUs as well as multicore CPUs via a PTX to LLVM (Low Level Virtual Machine) translator.
+The project is not in active development anymore.
+It only supports PTX up to version 3.1 (current version is 4.2).
```
Comment (psychocoderHPC) [inline doc/markdown/user/Introduction.md]: Could we update `4.2` to `5.0`

File: doc/markdown/user/Introduction.md [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+| OpenACC			| :white_check_mark:	| :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x:               | :white_check_mark: |
+| OpenCL			| :white_check_mark:	| :white_check_mark: | :x:               | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x:               | :white_check_mark: |
+| SYCL			| :white_check_mark:	| (:ballot_box_with_check:) | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | (:ballot_box_with_check:) | :white_check_mark: |
+| C++AMP			| :white_check_mark:	| :white_check_mark: | :white_check_mark: | (:ballot_box_with_check:) | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x:               | :white_check_mark: |
+| KOKKOS			| :white_check_mark:	| :white_check_mark: | :white_check_mark: | :white_check_mark: | :large_orange_diamond: | :white_check_mark: | :white_check_mark: | :x:               | :large_orange_diamond: |
```
Comment (psychocoderHPC) [inline doc/markdown/user/Introduction.md]: KOKKOS must get a check mark for `Heterogenity`. It runs on some architectures than alpaka.

Comment (psychocoderHPC) [inline doc/markdown/user/Introduction.md]: In KOKKOS it is also possible to select the target architecture as policy of the parallel_for or other entry points to a parallel algorithm. The default behaviour is selected at make time if no explicit target is selected.
The same is possible for memory places.

File: doc/markdown/user/Abstraction.md (hunk 1 above)
Comment (BenjaminW3) [inline doc/markdown/user/Abstraction.md]: done

File: doc/markdown/user/Introduction.md (hunk 3 above)
Comment (BenjaminW3) [inline doc/markdown/user/Introduction.md]: done

File: doc/markdown/user/abstraction/Element.md (hunk 2 above)
Comment (BenjaminW3) [inline doc/markdown/user/abstraction/Element.md]: done

File: doc/markdown/user/Introduction.md (hunk 4 above)
Comment (BenjaminW3) [inline doc/markdown/user/Introduction.md]: done

File: doc/markdown/user/implementation/mapping/x86.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+[Warps](../../abstraction/Abstraction.md) can handle branches with divergent control flows of multiple threads.
+There is no equivalent hardware unit in a CPU supporting this.
+Therefore, the warp level can not be utilized on CPUs leading to a one-to-one mapping of threads to warps which does not violate the rules of the abstraction.
+
+Block
+-----
+
+### One Block Per Node
+
+By combining all threads/PUs of all processors on the node into one block, the number of synchronizing and communicating threads is enlarged.
```
Comment (BenjaminW3) [inline doc/markdown/user/implementation/mapping/x86.md]: PUs had been used for `processing unit`. The sentence has been made clearer.

**Comment by BenjaminW3** [comment]
All comments up to now should have been handled.

**Comment by ax3l** [comment]
@BenjaminW3 if you rename "entry-point" files per directory, such as `Index.md` to `README.md` the github online file browser will automatically render them. might come handy, but will need to update the links to those.

**Pull request merged**